### PR TITLE
otelgrpc: add custom attributes to the stats handler

### PR DESCRIPTION
**Pull request by inigohu**
Fixes https://github.com/open-telemetry/opentelemetry-go-contrib/issues/3894

### CODE REVIEW

**Comment by inigohu** [comment]
The StatsHandler tests did not pass because there seems to be a race condition. The problem is that I moved the span creation from `TagRPC` to `HandleRPC`, but I think the `HandleRPC` stage is the correct place to start the Span. WDYT?

**Comment by inigohu** [comment]
Friendly ping @dashpole @hanyuancheung

**Comment by dashpole** [comment]
This looks like it does a lot more than just add a labeler.  Would you mind doing general refactoring in its own PR?

**Comment by inigohu** [comment]
> This looks like it does a lot more than just add a labeler. Would you mind doing general refactoring in its own PR?

Sure. I have simplified this PR by adding only the labeler.

File: instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go [hunk 1]
```diff
@@ -30,6 +30,16 @@ type gRPCContext struct {
 	record           bool
 }
 
+// AddAttrs adds attributes to the given context.
+func AddAttrs(ctx context.Context, attrs ...attribute.KeyValue) context.Context {
```
Comment (dashpole) [inline instrumentation/google.golang.org/grpc/otelgrpc/stats_handler.go:34]: nit: Consider `WithAttributes` to match the naming pattern for functions that add to the context.

**Review by dashpole — COMMENTED** [review]
can you add tests?

**Review by dashpole — COMMENTED** [review]
Can you test the span from the stats handler as well?

**Comment by inigohu** [comment]
@dashpole PTAL

**Review by dashpole — COMMENTED** [review]
I thought we were going to use a context-based pattern (similar to the otelhttp.Labeler) to support this?  Am I misremembering?

**Comment by inigohu** [comment]
>I thought we were going to use a context-based pattern (similar to the otelhttp.Labeler) to support this? Am I misremembering?

Yes, sorry, I forgot to comment this last change. My first idea was to follow the `otelhttp.Labeler` strategy using the context to pass the labels, but when I was adding the tests I realised that maybe it doesn't make much sense to do it this way, because unlike `http.Handler` and `grpc.UnaryInterceptors`, you can't chain different `grpc.StatsHandler` to add labels and pass them through the context.

That's why I changed the approach. The behaviour differs from that of `otelhttp.Labeler`. That one is request/context based tagging and this one is server based tagging.

WDYT @dashpole ?

**Comment by dashpole** [comment]
A few thoughts/ideas:

* If we use constant labels as you propose here, what is the use-case?  Are constant labels actually useful?
* Should the Labeler take a stats.RPCTagInfo or stats.ConnTagInfo as an argument?
* If we still used the Labeler approach, could users set the labeler in the context using an interceptor? (Not sure if interceptors run before the stats handler?)

**Comment by inigohu** [comment]
>If we use constant labels as you propose here, what is the use-case? Are constant labels actually useful?

Static attributes could be useful for tagging different servers running on the same service but on different ports, e.g. _public_, _internal, backoffice, legacy_...

>Should the Labeler take a stats.RPCTagInfo or stats.ConnTagInfo as an argument?

These attributes are already being tagged in the spans/metrics.

>If we still used the Labeler approach, could users set the labeler in the context using an interceptor? (Not sure if interceptors run before the stats handler?)

The `HandleRPC` of the `StatsHandler` is executed several times, but to summarise this would be the order:

HandleRPC(begin) -> Interceptors(pre) -> handler logic -> Interceptors(post) -> HandleRPC(end).

The problem is that despite modifying the context in an interceptor, [the interceptor does not propagate it outwards](https://github.com/grpc/grpc-go/blob/c04b085930ce33ee83cc3f92dbe7632031e127a9/server.go#L1379), so it never reaches [the end of the HandleRPC](https://github.com/grpc/grpc-go/blob/c04b085930ce33ee83cc3f92dbe7632031e127a9/server.go#L1246).

**Comment by jklipp** [comment]
Hi, I've ended up here after trying to figure out how to attach custom (static) labels to otelgrpc stats.  Just wanted to add our concrete use case if it's helpful to demonstrate the need.

We've got a central cluster that talks to the same service on multiple remote clusters, and I'd like to be able to label and differentiate remote clusters so I can examine their latency metrics separately.

From what I've been able to determine poking around the codebases, attaching a different client handler with a different set of static labels/attributes on it to each of our outbound grpc client connections would let us distinguish between the connections after everything rolls up to the client latency metric.

The only other way I can see to achieve that is to have different MeterProviders with different labels, but then I believe they would conflict on the metric names if I were to try and publish the metrics through the same exporter (the otel prometheus exporter in this case).

**Comment by dashpole** [comment]
I think this change is a good idea.  @hanyuancheung any thoughts?

**Review by dashpole — APPROVED** [review]
please add a changelog entry

File: CHANGELOG.md [hunk 2]
```diff
@@ -12,6 +12,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 
 - Support for stdoutlog exporter in `go.opentelemetry.io/contrib/config`. (#5850)
 - Add macOS ARM64 platform to the compatibility testing suite. (#5868)
+- Add custom attributes to the stats handler in `go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc`. (#5133)
```
Comment (dmathieu) [inline CHANGELOG.md:15]:
```suggestion
- Add the `WithAttributes` method to set custom attributes to the stats handler in `go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc`. (#5133)
```

**Review by dmathieu — APPROVED** [review]

File: instrumentation/google.golang.org/grpc/otelgrpc/config.go [hunk 3]
```diff
@@ -257,3 +258,16 @@ func (o spanStartOption) apply(c *config) {
...
+
+type attributesOption struct{ a []attribute.KeyValue }
+
+func (o attributesOption) apply(c *config) {
+	if o.a != nil {
+		c.Attributes = o.a
+	}
+}
+
+// WithAttributes returns an Option to add custom attributes to the spans and metrics.
```
Comment (pellared) [inline instrumentation/google.golang.org/grpc/otelgrpc/config.go:270]: I find adding attributes to both spans and metrics problematic. For spans attributes are added to give more context, but for metrics they are used also for aggregation.

Comment (pellared) [reply instrumentation/google.golang.org/grpc/otelgrpc/config.go:270]: @dashpole WDYT?

Comment (dashpole) [reply instrumentation/google.golang.org/grpc/otelgrpc/config.go:270]: right... People should probably use `WithSpanOptions(WithAttributes(...` to add span attributes, right?

Comment (pellared) [reply instrumentation/google.golang.org/grpc/otelgrpc/config.go:270]: I am against span options as it gives to much (changing span kind etc). See https://github.com/open-telemetry/opentelemetry-go-contrib/pull/5431. Maybe dedicated options to set attributes for each signal?

Comment (inigohu) [reply instrumentation/google.golang.org/grpc/otelgrpc/config.go]: @dashpole updated

**Review by dashpole — APPROVED** [review]

**Comment by zchee** [comment]
@dmathieu @dashpole Sorry to interrupt, but when will this change be released (merged)?

**Pull request merged**